File: go/isogram/isogram.go
package isogram

import "strings"

// IsIsogram determines whether word is isogram or not
func IsIsogram(word string) bool {
	m := map[byte]bool{}
	word = strings.ToLower(word)
	l := len(word)
	for i := 0; i < l; i++ {
		let := word[i]
		if let != ' ' && let != '-' {
			if m[let] {
				return false
			}
			m[let] = true
		}
	}
	return true
}

// TODO: find out why method below is faster
// IsIsogram accepts a phrase and returns if it is an isogram.
/*
func IsIsogram(phrase string) bool {
	the_end := len(phrase)
	var letterFlags uint32 = 0
	for i := 0; i < the_end; i++ {
		letter := phrase[i]
		// if A - Z
		if letter > 64 && letter < 91 {
			// make a - z
			letter += 32
		}
		// if a - z
		if letter > 96 && letter < 123 {
			if (letterFlags & (1 << (letter - 'a'))) != 0 {
				return false
			} else {
				letterFlags |= (1 << (letter - 'a'))
			}
		}
	}
	return true
}
*/

File: c/isogram/isogram.c
#include "isogram.h"
#include <stddef.h>

#define LEN 256
#define ALLOWED(c) (c == ' ' || c == '-')
#define IS_LETTER(c) (('a' <= c && c <= 'z') || ('A' <= c && c <= 'Z'))

char to_lower(char c);

bool is_isogram(const char phrase[])
{
	int abc[LEN];
	int i, c;
	int iso;

	if (phrase == NULL)
		return false;

	iso = true;	

	for (i = 0; i < LEN; ++i)
		abc[i] = 0;	

	for (i = 0; phrase[i] != '\0'; ++i) {
		c = to_lower(phrase[i]);	
		++abc[c];
		if (abc[c]>1 && ALLOWED(c)==false)
			iso = false;
	}
		
	return iso;
}

char to_lower(char c)
{
	int r;
	r = 'a' - 'A';
	if (IS_LETTER(c) == true) {
		if (c < 'a')
			return c + r;
		else if (c > 'z')
			return c - r;
	}
	return c;
}
